src/GraphQl/Resolver/Factory/ResolverFactory.php

```diff
@@ -28,16 +29,18 @@ public function __construct(
     ) {
     }
 
-    public function __invoke(?string $resourceClass = null, ?string $rootClass = null, ?Operation $operation = null): callable
+    public function __invoke(?string $resourceClass = null, ?string $rootClass = null, ?Operation $operation = null, ?PropertyMetadataFactoryInterface $propertyMetadataFactory = null): callable
     {
-        return function (?array $source, array $args, $context, ResolveInfo $info) use ($resourceClass, $rootClass, $operation) {
-            // Data already fetched and normalized (field or nested resource)
-            if ($body = $source[$info->fieldName] ?? null) {
-                return $body;
-            }
+        return function (?array $source, array $args, $context, ResolveInfo $info) use ($resourceClass, $rootClass, $operation, $propertyMetadataFactory) {
+            if (\array_key_exists($info->fieldName, $source ?? [])) {
+                $body = $source[$info->fieldName];
 
-            if (null === $resourceClass && \array_key_exists($info->fieldName, $source ?? [])) {
-                return $body;
+                $propertyMetadata = $rootClass ? $propertyMetadataFactory?->create($rootClass, $info->fieldName) : null;
+                $propertySchemaType = $propertyMetadata?->getSchema()['type'] ?? null;
+                // Data already fetched and normalized (field or nested resource)
+                if ($body || null === $resourceClass || $propertyMetadata && 'array' !== $propertySchemaType) {
```

Comment:
```suggestion
                if ($body || null === $resourceClass || 'array' !== $propertySchemaType) {
```

Reply:
obsolete because `isCollection` of builtInType could have been used.